### DISCUSSION TITLE
Vector VRL Script Compiler Performance and Architecture

### DISCUSSION THREAD

**Discussion by wmyre** [vectordotdev/vector#19221] [General]
How does Vector's compiler of VRL scripts function entirely? For example, does Vector perform some kind of caching of compiled scripts in its data directory for quicker reloads, and is this sort of compiler functionality documented anywhere? I do understand that vector builds its topology based on the entirety of the configuration, but it seems like the builders performance diminishes when it has to compile VRL scripts.

**Comment by jszwedko** [comment]
Hi @wmyre ,

VRL programs are "compiled" when Vector starts up into a tree structure that is walked at runtime to execute for each incoming event. This is not cached anywhere and so subsequent Vector starts will recompile the programs.

Does that answer your question?

**Reply by jszwedko** [reply]
Just discovered you'd created an issue from this discussion: https://github.com/vectordotdev/vector/issues/19222. I'll lock this discussion so we can focus the conversation onto the issue.
